Controller in crash loop

It's been about a while since I actually _ran_ the controller, so I'm unsure how long this has been going on. This evening, I see the controller going into a `CrashLoopBackOff`.

Logs consistently appear as follows:

```
$ kubectl logs controller-311122271-6ia9a -n catalog --previous
2016/12/16 22:49:02 Checking for existence of service-instance.catalog.k8s.io
2016/12/16 22:49:02 Found existing TPR service-instance.catalog.k8s.io
2016/12/16 22:49:02 Checking for existence of service-binding.catalog.k8s.io
2016/12/16 22:49:02 Found existing TPR service-binding.catalog.k8s.io
2016/12/16 22:49:02 Checking for existence of service-broker.catalog.k8s.io
2016/12/16 22:49:02 Found existing TPR service-broker.catalog.k8s.io
2016/12/16 22:49:02 Checking for existence of service-class.catalog.k8s.io
2016/12/16 22:49:02 Found existing TPR service-class.catalog.k8s.io
2016/12/16 22:49:02 Thirdparty: {TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:binding.steward.deis.io GenerateName: Namespace: SelfLink:/apis/extensions/v1beta1/thirdpartyresources/binding.steward.deis.io UID:aa630fe8-b829-11e6-af42-020d21ce6dc3 ResourceVersion:223297 Generation:0 CreationTimestamp:2016-12-02 00:52:57 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[heritage:steward] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName:} Description:A binding to a service instance that was provisioned through Steward Versions:[{Name:v1}]}
2016/12/16 22:49:02 Thirdparty: {TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:broker.steward.deis.io GenerateName: Namespace: SelfLink:/apis/extensions/v1beta1/thirdpartyresources/broker.steward.deis.io UID:aa5d8a91-b829-11e6-af42-020d21ce6dc3 ResourceVersion:223294 Generation:0 CreationTimestamp:2016-12-02 00:52:57 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[heritage:steward] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName:} Description:A service broker which Steward may use to provision and bind to services Versions:[{Name:v1}]}
2016/12/16 22:49:02 Thirdparty: {TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:instance.steward.deis.io GenerateName: Namespace: SelfLink:/apis/extensions/v1beta1/thirdpartyresources/instance.steward.deis.io UID:aa60a361-b829-11e6-af42-020d21ce6dc3 ResourceVersion:223296 Generation:0 CreationTimestamp:2016-12-02 00:52:57 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[heritage:steward] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName:} Description:A service instance that Steward will provision through an underlying service broker Versions:[{Name:v1}]}
2016/12/16 22:49:02 Thirdparty: {TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:service-binding.catalog.k8s.io GenerateName: Namespace: SelfLink:/apis/extensions/v1beta1/thirdpartyresources/service-binding.catalog.k8s.io UID:9c00f592-c3e1-11e6-af42-020d21ce6dc3 ResourceVersion:2644802 Generation:0 CreationTimestamp:2016-12-16 22:47:23 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName:} Description:A Service Binding representation, creates a Service Binding Versions:[{Name:v1alpha1}]}
2016/12/16 22:49:02 Thirdparty: {TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:service-broker.catalog.k8s.io GenerateName: Namespace: SelfLink:/apis/extensions/v1beta1/thirdpartyresources/service-broker.catalog.k8s.io UID:9bec73f7-c3e1-11e6-af42-020d21ce6dc3 ResourceVersion:2644798 Generation:0 CreationTimestamp:2016-12-16 22:47:23 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName:} Description:A Service Broker representation. Adds a service broker and fetches its catalog Versions:[{Name:v1alpha1}]}
2016/12/16 22:49:02 Thirdparty: {TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:service-class.catalog.k8s.io GenerateName: Namespace: SelfLink:/apis/extensions/v1beta1/thirdpartyresources/service-class.catalog.k8s.io UID:9bed90e6-c3e1-11e6-af42-020d21ce6dc3 ResourceVersion:2644799 Generation:0 CreationTimestamp:2016-12-16 22:47:23 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName:} Description:A Service Type representation. Something that a customer can instantiate Versions:[{Name:v1alpha1}]}
2016/12/16 22:49:02 Thirdparty: {TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:service-class.steward.deis.io GenerateName: Namespace: SelfLink:/apis/extensions/v1beta1/thirdpartyresources/service-class.steward.deis.io UID:aa5edba0-b829-11e6-af42-020d21ce6dc3 ResourceVersion:223295 Generation:0 CreationTimestamp:2016-12-02 00:52:57 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[heritage:steward] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName:} Description:A type of service that Steward can provision through an underlying service broker Versions:[{Name:v1}]}
2016/12/16 22:49:02 Thirdparty: {TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:service-instance.catalog.k8s.io GenerateName: Namespace: SelfLink:/apis/extensions/v1beta1/thirdpartyresources/service-instance.catalog.k8s.io UID:9bef09e0-c3e1-11e6-af42-020d21ce6dc3 ResourceVersion:2644800 Generation:0 CreationTimestamp:2016-12-16 22:47:23 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName:} Description:A Service Instance representation, creates a Service Instance Versions:[{Name:v1alpha1}]}
2016/12/16 22:49:02 initCluster
2016/12/16 22:49:02 Checking resource type ServiceInstance for readiness for listing
2016/12/16 22:49:02 Successful list for ServiceInstance, continuing
2016/12/16 22:49:02 Checking resource type ServiceBinding for readiness for listing
2016/12/16 22:49:02 Successful list for ServiceBinding, continuing
2016/12/16 22:49:02 Checking resource type ServiceBroker for readiness for listing
2016/12/16 22:49:02 Successful list for ServiceBroker, continuing
2016/12/16 22:49:02 Checking resource type ServiceClass for readiness for listing
2016/12/16 22:49:02 Successful list for ServiceClass, continuing
2016/12/16 22:49:02 Starting to watch for new Service Brokers
2016/12/16 22:49:02 Starting to watch for new Service Instances
2016/12/16 22:49:02 Starting to watch for new Service Bindings
2016/12/16 22:49:02 Server started on port 10000
2016/12/16 22:49:02 Entering watch loop
2016/12/16 22:49:02 Entering watch loop
2016/12/16 22:49:02 Entering watch loop
```

As you can see, there is no obvious reason for the failure in the logs.

Note that I've tried this on two different clusters with the same results:

- minikube w/ k8s 1.4.5
- kube-aws w/ k8s 1.4.6

Is anyone else running into this?

I guess I did a good refactor, because the problem was perfectly preserved. :wink:

There's more info to go on now though:

```
W0112 19:26:03.398842       1 client_config.go:438] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
I0112 19:26:03.399500       1 watcher.go:170] Checking for existence of service-broker.catalog.k8s.io
E0112 19:26:03.426676       1 watcher.go:173] Found existing TPR service-broker.catalog.k8s.io
I0112 19:26:03.426858       1 watcher.go:170] Checking for existence of service-class.catalog.k8s.io
E0112 19:26:03.428249       1 watcher.go:173] Found existing TPR service-class.catalog.k8s.io
I0112 19:26:03.428291       1 watcher.go:170] Checking for existence of service-instance.catalog.k8s.io
E0112 19:26:03.429687       1 watcher.go:173] Found existing TPR service-instance.catalog.k8s.io
I0112 19:26:03.429755       1 watcher.go:170] Checking for existence of service-binding.catalog.k8s.io
E0112 19:26:03.431114       1 watcher.go:173] Found existing TPR service-binding.catalog.k8s.io
I0112 19:26:03.432494       1 watcher.go:201] Thirdparty: {TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:service-binding.catalog.k8s.io GenerateName: Namespace: SelfLink:/apis/extensions/v1beta1/thirdpartyresources/service-binding.catalog.k8s.io UID:d437a07d-d8fc-11e6-b46e-8614ef398da2 ResourceVersion:4616 Generation:0 CreationTimestamp:2017-01-12 19:25:08 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName:} Description:A Service Binding representation, creates a Service Binding Versions:[{Name:v1alpha1}]}
I0112 19:26:03.432565       1 watcher.go:201] Thirdparty: {TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:service-broker.catalog.k8s.io GenerateName: Namespace: SelfLink:/apis/extensions/v1beta1/thirdpartyresources/service-broker.catalog.k8s.io UID:d4385546-d8fc-11e6-b46e-8614ef398da2 ResourceVersion:4617 Generation:0 CreationTimestamp:2017-01-12 19:25:08 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName:} Description:A Service Broker representation. Adds a service broker and fetches its catalog Versions:[{Name:v1alpha1}]}
I0112 19:26:03.432585       1 watcher.go:201] Thirdparty: {TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:service-class.catalog.k8s.io GenerateName: Namespace: SelfLink:/apis/extensions/v1beta1/thirdpartyresources/service-class.catalog.k8s.io UID:d438df4e-d8fc-11e6-b46e-8614ef398da2 ResourceVersion:4618 Generation:0 CreationTimestamp:2017-01-12 19:25:08 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName:} Description:A Service Type representation. Something that a customer can instantiate Versions:[{Name:v1alpha1}]}
I0112 19:26:03.432652       1 watcher.go:201] Thirdparty: {TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:service-instance.catalog.k8s.io GenerateName: Namespace: SelfLink:/apis/extensions/v1beta1/thirdpartyresources/service-instance.catalog.k8s.io UID:d45218aa-d8fc-11e6-b46e-8614ef398da2 ResourceVersion:4619 Generation:0 CreationTimestamp:2017-01-12 19:25:08 +0000 UTC DeletionTimestamp:<nil> DeletionGracePeriodSeconds:<nil> Labels:map[] Annotations:map[] OwnerReferences:[] Finalizers:[] ClusterName:} Description:A Service Instance representation, creates a Service Instance Versions:[{Name:v1alpha1}]}
I0112 19:26:03.432670       1 watcher.go:207] initCluster
I0112 19:26:03.432677       1 watcher.go:215] Checking resource type ServiceInstance for readiness for listing
I0112 19:26:03.433890       1 watcher.go:220] Successful list for ServiceInstance, continuing
I0112 19:26:03.433918       1 watcher.go:215] Checking resource type ServiceBinding for readiness for listing
I0112 19:26:03.434938       1 watcher.go:220] Successful list for ServiceBinding, continuing
I0112 19:26:03.434967       1 watcher.go:215] Checking resource type ServiceBroker for readiness for listing
I0112 19:26:03.435887       1 watcher.go:220] Successful list for ServiceBroker, continuing
I0112 19:26:03.435916       1 watcher.go:215] Checking resource type ServiceClass for readiness for listing
I0112 19:26:03.436901       1 watcher.go:220] Successful list for ServiceClass, continuing
I0112 19:26:03.437410       1 controller.go:70] Starting to watch for new Service Brokers
I0112 19:26:03.437436       1 controller.go:77] Starting to watch for new Service Instances
I0112 19:26:03.438505       1 watcher.go:323] Entering watch loop
I0112 19:26:03.437439       1 controller.go:84] Starting to watch for new Service Bindings
panic: unreachable

goroutine 1 [running]:
panic(0x1452280, 0xc4204a9fb0)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/kubernetes-incubator/service-catalog/cmd/controller/app.Run(0xc4203c29c0, 0x171b5c0, 0xc42055fe98)
	/go/src/github.com/kubernetes-incubator/service-catalog/cmd/controller/app/controller.go:86 +0x5d6
main.main()
	/go/src/github.com/kubernetes-incubator/service-catalog/cmd/controller/controller.go:50 +0xc1
```

`panic: unreachable` is where the controller's `main()` function exits. I read this as a pretty strong indication that the controller's `Run()` function isn't blocking like it should. I will look into this further.

Assigning to myself.

--- a/cmd/controller/app/controller.go
+++ b/cmd/controller/app/controller.go
@@ -71,17 +71,15 @@ func Run(s *options.ControllerServer) error {
 		panic(fmt.Sprintf("Error creating server [%s]...", err.Error()))
 	}
 
-	go func() {
-		mux := http.NewServeMux()
-		healthz.InstallHandler(mux)
-		server := &http.Server{
-			Addr:    net.JoinHostPort(s.Address, strconv.Itoa(int(s.Port))),
-			Handler: mux,
-		}
-		glog.Fatal(server.ListenAndServe())
-	}()
-
 	c.Run()
 
-	panic("unreachable")
+	mux := http.NewServeMux()
+	healthz.InstallHandler(mux)
+	server := &http.Server{
+		Addr:    net.JoinHostPort(s.Address, strconv.Itoa(int(s.Port))),
+		Handler: mux,
+	}
+	glog.Fatal(server.ListenAndServe())
+
+	return nil
 }

